LIP-36: align proposal with latest implementation (#125)
lidofinance/lido-improvement-proposals: LIPS/lip-36.md

Reviewer: ```suggestion
TwoCrypto-NG pools have no pool-level pause, but are not fully immutable: the Curve factory admin can ramp A/gamma and update fee/rebalancing/oracle parameters. The Curve factory address, implementation address, math/view implementations, factory admin, fee receiver, initial price, A/gamma, fees, and rebalancing parameters will be published before the on-chain vote.
```

Author: Fixed in [bcaa05e](https://github.com/lidofinance/lido-improvement-proposals/pull/125/commits/bcaa05e61a3961cb9f50081e3dc8f636bd786964).

Reviewer: ```suggestion
The pool may launch with zero liquidity because TwoCrypto-NG initializes `price_scale`, `price_oracle`, and `last_prices` from `initial_price`. NEST does not use a TVL-based bootstrap bypass. The `price_scale()` divergence gate is active from the first deposit.
```

Reviewer: ```suggestion
Zero-seed launch is mechanically valid, but it is not risk-free. It relies on a correct `initial_price`, correct Curve implementation provenance, and the zero-supply pool-state check. Controlled seed liquidity remains an alternative launch-risk mitigation, not a mechanical contradiction.
```

Reviewer: I think the Curve-side protection model might be changed before this is finalized.

The current implementation gates `addLiquidity()` on Curve `price_oracle()` and bypasses that gate while pool TVL is below `poolBootstrapMinTvlUsd`. That doesn't look like the best check for TwoCrypto-NG deposits.

In TwoCrypto-NG, `price_oracle()` is an EMA/signal. The deposit math rather uses `price_scale()` as the accounting price:

- `add_liquidity()` computes the new invariant using current balances and `cached_price_scale`.
- `price_oracle()` can be stale by design and is updated from previous state prices.
- A zero-liquidity pool is not a reason to bypass price checks, because `price_scale`, `price_oracle`, and `last_prices` are initialized from `initial_price` at deployment.

So the TVL-based bootstrap bypass should be removed. The first-liquidity path should instead rely on a correct `initial_price` and an always-on `price_scale()` vs `OracleRouter` gate.

Suggested implementation direction:

1. Replace the pool EMA gate with an always-on `price_scale()` gate:

```solidity
uint256 poolScaleLdoPerStEth = Math.mulDiv(
    CURVE_POOL_AND_TOKEN.price_scale(),
    PRICE_UNIT,
    WSTETH.stEthPerToken()
);

divergenceBps = _absDivergenceBps(poolScaleLdoPerStEth, oracleLdoPerStEth);

status = divergenceBps > poolPriceDivergenceToleranceBps
    ? AddLiquidityStatus.PoolPriceDivergenceTooHigh
    : AddLiquidityStatus.Eligible;
```

2. Delete the TVL bootstrap mechanism:

- `poolBootstrapMinTvlUsd`
- `setPoolBootstrapMinTvlUsd`
- `PoolBootstrapMinTvlUsdSet`
- `InvalidPoolBootstrapMinTvlUsd`
- `MAX_POOL_BOOTSTRAP_MIN_TVL_USD`
- the `InitParams.poolBootstrapMinTvlUsd` field
- the `_poolTvlUsd(...) >= poolBootstrapMinTvlUsd` condition

3. Add an LP entry-value guard for nonzero-supply pools. A same-tx `calc_token_amount()` quote alone is mostly a sanity check, because it quotes the same state the deposit will use. The more meaningful guard is: quote the LP tokens, value the resulting LP share against oracle-priced pool NAV, and reject if the LP position would be worth materially less than the oracle value of the deposited assets.

Example shape:

```solidity
uint256[2] memory amounts = [ldoAmount, wstEthAmount];
uint256 quotedLp = CURVE_POOL_AND_TOKEN.calc_token_amount(amounts, true);

uint256 lpEntryValueUsd = Math.mulDiv(
    quotedLp,
    poolNavUsd + depositValueUsd,
    poolLpSupply + quotedLp
);

uint256 minLpEntryValueUsd = Math.mulDiv(
    depositValueUsd,
    MAX_BASIS_POINTS - maxLpEntryLossBps,
    MAX_BASIS_POINTS,
    Math.Rounding.Up
);

if (lpEntryValueUsd < minLpEntryValueUsd) {
    revert LpEntryValueTooLow(lpEntryValueUsd, minLpEntryValueUsd, maxLpEntryLossBps);
}
```

4. Treat zero-supply pools separately. If LP supply is zero, NEST becomes the sole LP and there are no existing LPs to donate value to. Require pool NAV to be zero and use `min_mint_amount = 1` for that first deposit, because Curve's external `calc_token_amount()` view path is not reliable for an empty pool.

5. Extend `ICurvePool` with the required TwoCrypto-NG views:

```solidity
function price_scale() external view returns (uint256);
function calc_token_amount(uint256[2] calldata amounts, bool deposit) external view returns (uint256);
```

This also means the LIP might be updated:

- remove the claim that bootstrap deposits run without protection because EMA needs TVL to converge;
- state that `price_scale()` is the load-bearing guard because it is the price used by liquidity accounting;
- state that `price_oracle()` is optional defense-in-depth/monitoring, not the primary deposit valuation guard;
- remove `poolBootstrapMinTvlUsd` from launch parameters;
- add `maxLpEntryLossBps`;
- document that zero-seed launch depends on correct `initial_price` and Curve deployment provenance.

—

See a possible patch shape for the impl:
[nest-curve-better-guard.patch](https://github.com/user-attachments/files/29480169/nest-curve-better-guard.patch)

**Transition behaviors:**

- **Surplus to deficit**: when the reserve outpaces revenue, `delta` goes negative and `budgetUSD` falls, possibly below zero; a negative budget spends nothing and is held in storage until later surplus lifts it back above zero.
- **Deficit to surplus**: as revenue recovers, positive `delta` rebuilds `budgetUSD` and buybacks resume automatically once it crosses back above zero.
- **Share and rate changes**: `setSurplusShareBP` and `setReserveDailyRateUSD` checkpoint the budget at the old value first, so a new share or rate applies only to revenue and days after the call. Budget already banked in `budgetUSD` is unaffected; there is no retroactive reapplication and no reset lever.

---

### Liquidity Provisioning

The `BuybackExecutor` receives settled LDO from CoW Swap and holds the stETH kept aside for pairing. Liquidity provisioning runs as a separate permissionless flow after trade settlement.

`addLiquidity()` flow:

1. Revert if the system is not in LP mode.
2. Read LDO and stETH balances; revert if either is zero.
3. Convert both to USD via the OracleRouter; revert if any price feed is stale or unavailable.

> 4. Check the pool price divergence gate: revert if the pool EMA diverges from the oracle ratio beyond `poolPriceDivergenceToleranceBps` (bypassed during bootstrap; see [Pool Skew Protection](#pool-skew-protection)).

Reviewer: ```suggestion
4. Check the pool price divergence gate: revert if Curve `price_scale()`, converted from LDO/wstETH to LDO/stETH, diverges from the OracleRouter ratio beyond `poolPriceDivergenceToleranceBps`. This gate is always on.
```

### System Architecture

NEST is implemented as two cooperating but independently operable contracts — the BuybackAllocator (budget allocation) and the BuybackExecutor (trade execution and liquidity provisioning). They utilize existing contracts like Stonks v2 and OracleRouter, as well as the governance infrastructure.

The system operates in two governance-controlled modes:

1. **LP mode** — acquired LDO is paired with wstETH and deposited into a Curve LDO/wstETH TwoCrypto-NG pool; LP tokens are retained by the BuybackExecutor as DAO-owned liquidity.
2. **Treasury-only mode** — acquired LDO is delivered directly to the Aragon Agent treasury.

> The active mode is derived from the settlement receiver of the executor's active Stonks instance: when the receiver is the executor, LP mode is active; when it is the treasury, treasury-only mode is active. Switching between modes means pointing the executor at a different Stonks instance via `setStonksAndOperatingMode` and requires a full DAO vote.

Reviewer: `setStonksAndOperatingMode()` can abandon a live tracked order: BuybackExecutor.sol. 

Prob the LIP should state that switching modes must only happen after no live order remains, or code should enforce it.

This LIP proposes deploying NEST in **LP mode as the initial operating configuration**. Treasury-only mode is available as a governance-controlled fallback that could be enabled by a subsequent DAO vote without redeploying the core contracts.

### New Contracts

#### `BuybackAllocator`

> Central budget engine of the system. Holds the funded stETH and a set of revenue source addresses, aggregates their cumulative revenue, evaluates eligibility, enforces daily and yearly caps, and pays out stETH to a single executor via `allocate()`. Activated once via `activate()`. Exposes `resetAccounting()` and governance setters. Order creation, trading, and Stonks/Order management live in the `BuybackExecutor`.

Reviewer: we don't have `resetAccounting` in the current impl iirc

### Trade Execution

Execution is two independent permissionless calls:

1. **`allocate()` on the `BuybackAllocator`** computes the spendable USD under the surplus model and the caps, converts it to stETH, transfers it to the executor, and calls the executor's allocation hook. If nothing is eligible it emits `AllocationSkipped` and returns without moving funds.
2. **`placeOrder()` on the `BuybackExecutor`** sells the stETH forwarded to Stonks (sized between configured min/max order bounds) for LDO via a CoW Swap order, with `minBuyAmount` computed onchain from `Stonks.estimateTradeOutput`.

On each allocation the executor sets aside stETH still committed in the pipeline and forwards the rest to Stonks — the full amount in treasury-only mode, half in LP mode, keeping the other half to pair as liquidity. Cadence is governed by the clock-based reserve and the fixed cap windows, and by a single-live-order guard on the executor.

#### Eligibility Gates

> **`allocate()` skips (non-reverting; emit `AllocationSkipped`):** not activated, oracle price unavailable, stETH price below floor, no available budget (allowance or a cap exhausted), payout below the per-call minimum.

Reviewer: when not activated it reverts though (according to `whenActivated` modifier in the current impl)

**`placeOrder()` preconditions (revert on failure):** executor pause, no live tracked order, Stonks balance at or above the minimum order size.

#### Surplus Mechanics

Revenue surplus accumulates cumulatively against a baseline set at activation and a clock-based reserve. A configured share of the all-time surplus is the all-time spending allowance; buybacks are only allowed while that allowance exceeds what has already been spent.

```
surplus      = Σ cumulativeRevenueUSD − revenueBaselineUSD − reserveCumulative
allowance    = max(0, surplus) × surplusShareBP / 10000

> spendableUSD = allowance − totalSpentUSD

Reviewer: The contract is incremental: it snapshots `budgetDelta = share·(Δrevenue − Δreserve)` into a signed `budgetUSD` each checkpoint, moves the baseline forward, and spends `clamp(budgetUSD)`. 

There is no stored `revenueBaselineUSD`, no `reserveCumulative`, and no `totalSpentUSD` accumulator. The two are equivalent only while the surplus share is constant; the moment it changes they diverge (mind the gap about ordering there, noted). 

Would switch to the actual model — a signed running budget with per‑checkpoint share application — or the formula will mislead anyone reasoning about share changes or deficit memory.

```

Execution requires all of the following:

1. The allocator is activated.
2. The oracle returns a stETH/USD price.
3. `stEthPriceUSD ≥ minStEthPriceUSD` — the price floor, inactive when `0`.
4. Spendable budget remains after the allowance, cap, and balance clamps.

`reserveCumulative` grows by `reserveDailyRateUSD` for each elapsed day since activation. `spendableUSD` is clamped by the unused daily and yearly cap windows, the held stETH balance, and a per-call minimum. In LP mode the executor sells half of the forwarded stETH for LDO and pairs the other half as wstETH; in treasury-only mode the full amount is swapped for LDO.

**Transition behaviors:**

- **Surplus to deficit**: when the reserve outpaces revenue, surplus shrinks and `spendableUSD` falls to zero; buybacks pause until new revenue rebuilds surplus.
- **Deficit to surplus**: as revenue recovers, surplus grows back above the reserve and buybacks resume automatically.

> - **Share changes**: `surplusShareBP` is applied to the entire surplus history on every call, so a change takes effect across accrued surplus; pair it with `resetAccounting()` to scope it to future revenue only.

Reviewer: Code does the opposite: `setSurplusShareBP()` checkpoints first at the old share, then applies the new share only to future deltas.

Prob this LIP should describe the signed `budgetUSD` delta model: `budgetUSD += (revenueDelta - reserveAccrued) * surplusShareBP / 10000`, then spends decrement `budgetUSD`.

#### Pool Skew Protection

Protection scheme:

> 1. **Pool price divergence gate** compares the Curve pool's internal EMA price from `price_oracle()` against the OracleRouter LDO/wstETH price. If divergence exceeds `poolPriceDivergenceToleranceBps`, the deposit reverts. This is a structural check independent of deposit size.

Reviewer: This omits the bootstrap bypass. 

The Pool Skew Protection section states the gate reverts on divergence as an always‑on structural check, but the code bypasses it entirely below `poolBootstrapMinTvlUsd` (acknowledged only in the parameter table). 

During bootstrap there is therefore no divergence gate and no slippage guard — the riskiest window is the least protected. The section should state this plainly, and reconcile with the gate‑bypass parameter. (Also a unit nit: the gate compares LDO/stETH, converting the EMA via stEthPerToken, not LDO/wstETH as written). So once again this goes back to `1` for slippage min param.

> 2. Because the pool EMA moves across blocks it resists in-block manipulation, so the Curve deposit relies on this gate for sandwich resistance rather than a per-deposit slippage parameter.

Reviewer: not sure about this still (would seek for more confirmations somehow) — noted this in the impl PR too

#### LP Token Management

| Function                                                     | Access         | Description                                                                       |
| ------------------------------------------------------------ | -------------- | --------------------------------------------------------------------------------- |
| `removeLiquidityAndRecoverToTreasury(uint256 lpAmount, ...)` | `MANAGER_ROLE` | Burns LP tokens, unwraps wstETH, sends LDO and stETH to the Aragon Agent treasury |
| `recoverERC20(address token, uint256 amount)`                | `MANAGER_ROLE` | Recovers assets (including LP tokens) to the Aragon Agent treasury                |

---

### Off-chain Keeper

A keeper bot polls the BuybackAllocator and BuybackExecutor via view functions on a schedule (e.g., every 15 minutes):

1. Call `convertPendingRevenueToUSD()` to settle pending staking revenue when the oracle is healthy.
2. Check `spendable()` → call `allocate()` if eligible.
3. Check `getPlacementStatus()` → call `placeOrder()` if eligible.
4. Check `canAddLiquidity()` → call `addLiquidity()` if eligible.

> All functions are permissionless and idempotent; failed calls are retried on the next cycle. The keeper is integrated into the existing `MotionEnacterBot`, reusing existing infrastructure.

Reviewer: `allocate()`, `placeOrder()`, and `addLiquidity()` are permissionless but not idempotent as they move funds or create orders. 

Prob replace with "permissionless and retryable when preconditions are unmet".

#### Pool Skew Protection

> - **Pool price divergence gate.** Compares the Curve pool's internal EMA price from `price_oracle()`, which reports LDO per wstETH and is converted to an LDO/stETH ratio via `wstETH.stEthPerToken()`, against the OracleRouter's LDO/stETH price. If divergence exceeds `poolPriceDivergenceToleranceBps`, the deposit reverts. This is a structural check independent of deposit size.

Reviewer: ```suggestion
- **Mandatory price-scale gate.** NEST compares Curve `price_scale()` against `OracleRouter` because TwoCrypto-NG values liquidity deposits against `price_scale`, not `price_oracle()`. A deposit reverts if divergence exceeds `poolPriceDivergenceToleranceBps`.
```

> - **No per-deposit slippage floor.** Curve values an imbalanced deposit at the pool's internal `price_scale`, which moves only gradually and whose EMA updates at most once per block (the state price is capped at `2 × price_scale`), so a deposit cannot be sandwiched in-block by moving the spot price. This in-block resistance is intrinsic to Curve, independent of NEST's gate. The `add_liquidity` call itself passes a minimum-LP-out of `1` (no effective slippage floor), so the divergence gate is the _sole_ guard against the risk Curve's in-block resistance does not cover: depositing while the pool's price has drifted from the external market.

Reviewer: ```suggestion
- **LP entry-value guard.** For nonzero-supply pools, NEST quotes `calc_token_amount(amounts, true)`, values the quoted LP share against oracle-priced pool NAV, and reverts if the LP position would be worth less than the deposited oracle value by more than `maxLpEntryLossBps`. For zero-supply pools, NEST requires zero pool NAV and accepts becoming the sole LP.
```

> - **The bootstrap window has neither protection.** While pool TVL is below `poolBootstrapMinTvlUsd` the divergence gate is bypassed (see [Pool Bootstrap](#pool-bootstrap)), and the deposit already carries no slippage floor, so a bootstrap deposit runs with no divergence gate and no slippage guard. This is the least-protected window, accepted because a shallow pool's stale EMA cannot converge without the very deposits the gate would otherwise block. The gate evaluates pre-deposit pool TVL, so `poolBootstrapMinTvlUsd` together with the per-call `maxDepositValueUsd` bounds the unguarded exposure: it is the threshold plus at most one capped deposit, after which the gate engages.

Reviewer: ```suggestion
- **No TVL bootstrap bypass.** The first-liquidity risk is handled by the always-on `price_scale()` gate and the zero-supply pool-state check. If market price moves away from the configured `initial_price`, deposits halt until governance updates deployment/configuration.
```

| Simple revenue percentage split      | Does not enforce surplus discipline; can activate with no actual treasury surplus       |

Expired and unfilled orders are retried through a bounded recovery flow rather than silently skipped: an expired order's stETH is swept back to Stonks and re-sold by the permissionless `placeOrder()` without additional funding. Because wstETH is minted only at deposit time, partial fills leave no wstETH overhang to clean up.

---

## Security Considerations

### Oracle Manipulation

Price feeds (ETH/USD, stETH/USD via two-hop, LDO/ETH) are sourced from Chainlink via the shared OracleRouter. All feeds are validated for staleness and answer validity independently. The daily cap bounds maximum exposure per day to $50k regardless of oracle accuracy. The OracleRouter can be paused independently for rapid incident response.

### Permissionless Execution Risk

`allocate()` and `placeOrder()` are callable by anyone. Budgets and `minBuyAmount` use on-chain oracle prices at call time; slippage protection is the Stonks-side margin. Daily and yearly caps enforce hard spending bounds regardless of oracle precision.

> ### Revenue Reporting Integrity

Reviewer: would also highlight that revenue source must be:
- monolithic
- always-alive
- fully settled

the latter is important upon activation or addition of a new one, e.g.

also liveness/accounting risk is handled with try/catch, but once again, current impl makes revertable source unremovable

> `StakingRevenueSource`'s rebase callback is authorized live against `LidoLocator.postTokenRebaseReceiver()`, and replays are rejected by a strictly-increasing report timestamp. USD conversion is deferred to a retryable permissionless call, so a price-feed outage defers rather than loses revenue.

Reviewer: The LIP must add an explicit deployment dependency: NEST requires a new/upgraded TokenRateNotifier that (a) ERC‑165‑detects the args‑bearing observer and (b) forwards the full rebase payload. 

Now LIP assumes that it's already here.
